src/cart/crt_swim.c

```diff
@@ -584,6 +629,30 @@ static void crt_swim_cli_cb(const struct crt_cb_info *cb_info)
 			  DP_RC(rpc_out->rc), DP_RC(rc));
 
 out:
+	if (rpc_out->excl_grp_ver > 0) {
+		D_RWLOCK_RDLOCK(&grp_priv->gp_rwlock);
+		if (grp_priv->gp_membs_ver > 0 && rpc_out->excl_grp_ver > grp_priv->gp_membs_ver) {
+			struct crt_swim_membs	*csm = &grp_priv->gp_membs_swim;
+			struct crt_swim_target	*cst;
+			uint64_t		 incarnation = 0;
+
+			/*
+			 * I'm excluded. TODO: Strictly speaking, we should
+			 * compare rpc_out->excl_grp_ver with the group version
+			 * at which we joined the system.
+			 */
+			D_WARN("excluded in group version %u (self %u)\n", rpc_out->excl_grp_ver,
+			       grp_priv->gp_membs_ver);
+			crt_swim_csm_lock(csm);
+			cst = crt_swim_membs_find(csm, self_id);
+			if (cst != NULL)
+				incarnation = cst->cst_state.sms_incarnation;
+			crt_swim_csm_unlock(csm);
+			crt_trigger_event_cbs(self_id, incarnation, CRT_EVS_GRPMOD, CRT_EVT_DEAD);
```

Comment:
so what is supposed to happen in the end upon self eviction? does self propagate eventually some new rejoin and update everyone that it is alive afterall?

Reply:
With this patch, the engine will commit suicide and remain stopped until the administrator starts it again. In the future, the control plane may decide whether to start this engine again automatically. Rejoining without restarting would be too complex to implement in my opinion; it's simpler to always stop the engine.